File: Grafos/exemplo-busca-em-profundidade-lista-de-adjacencias.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <limits>
using namespace std;
/**
 * A lista de adjacências representa o grafo em espaço O(|V|+|E|)
 * e possibilita busca em largura ou profundidade em tempo O(|V|+|E|)
 */

void dfs(vector<vector<int>>& g, int v){
    const int white = 0; //vértice não processado
    const int grey = 1; //vértice em processamento
    const int black = 2; // vértice processado
    vector<int> color(g.size(),white); //cor
    vector<int> pai(g.size(),-1); //informação do pai
    color[v]=grey;
    stack<int> s;
    s.push(v);
    while(!s.empty()){
        int u = s.top(); // retira o nó da pilha
        s.pop();
        cout << "Nó visitado = " << u << endl;
        cout << "Pai do nó = " << pai[u] << endl;
        for(size_t j=0;j<g[u].size();j++){ //percorre a lista de adjacências
            int w = g[u][j]; //retira o nó adjacente
            if(color[w]==white){ //se o nó ainda não foi visitado, inclue ele na lista
                color[w] = grey; // marca o nó como em processamento
                pai[w] = u; //atualiza o pai
                s.push(w); //coloca o nó na pilha
            }
        }
        color[u] = black;       
    }
}





int main(){
    vector<vector<int>> g;
    g.resize(5);
    g[0] = {1,4};
    g[1] = {2,3};
    g[2] = {3,4};
    g[3] = {3,3};
    g[4] = {0,1};
    dfs(g,4);
    return 0;
}

File: Grafos/_exemplo_busca_em_profundidade_lista_de_adjacencias.java
package grafos;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Deque;
import java.util.List;

public class _exemplo_busca_em_profundidade_lista_de_adjacencias {
	static void dfs(	List<List<Integer>> g, int v){
	    int white = 0; //vértice não processado
	    int grey = 1; //vértice em processamento
	    int black = 2; // vértice processado
	    List<Integer> color = new ArrayList<Integer>(Collections.nCopies(g.size(), white));
		List<Integer> pai = new ArrayList<Integer>(Collections.nCopies(g.size(), -1));
		
	    color.set(v, grey);
	    Deque<Integer> s = new ArrayDeque<Integer>();
	    s.push(v);
	    while(!s.isEmpty()){
	        int u = s.peek(); // retira o nó da pilha
	        s.pop();
	        System.out.println("Nó visitado = "+u);
	        System.out.println("Pai do nó = "+pai.get(u));
	        for(int j=0;j<g.get(u).size();j++){ //percorre a lista de adjacências
	            int w = g.get(u).get(j); //retira o nó adjacente
	            if(color.get(w)==white){ //se o nó ainda não foi visitado, inclue ele na lista
	               color.set(w, grey); // marca o nó como em processamento
	                pai.set(w, u); //atualiza o pai
	                s.push(w); //coloca o nó na pilha
	            }
	        }
	        color.set(u, black);       
	    }
	}

	public static void main(String[] args) {
		List<List<Integer>> g = new ArrayList<List<Integer>>();
		g.add(new ArrayList<Integer>());
		g.add(new ArrayList<Integer>());
		g.add(new ArrayList<Integer>());
		g.add(new ArrayList<Integer>());
		g.add(new ArrayList<Integer>());
		g.get(0).add(1);
		g.get(0).add(4);
		g.get(1).add(2);
		g.get(1).add(3);
		g.get(2).add(3);
		g.get(2).add(4);
		g.get(3).add(3);
		g.get(3).add(3);
		g.get(4).add(0);
		g.get(4).add(1);

		dfs(g, 4);
		
	}

}
